keps/sig-node/3759-sidecar-containers/README.md

```diff
@@ -0,0 +1,1235 @@
+
+### Problems: service mesh
+
+Service mesh presents a similar problem: you want the service mesh container to
+be running and ready before other containers start, so that any inbound/outbound
+connections that a container can initiate goes through the service mesh.
+
+A similar problem happens for shutdown: if the service mesh container is
+terminated prior to the other containers, outgoing traffic from other apps will
+be blackholed or not use the service mesh.
+
+However, as none of these are possible to guarantee, most service meshes (like
+Linkerd and Istio), need to do several hacks to have the basic functionality.
+
+Some service meshes depend on secret (like a certificate) downloaded by other
+Init container to establish secure communication between services. This makes
+the problem of ordering of sidecar and Init containers harder.
+
+### Problems: configuration / secrets
+
+Some pods use init containers to pull down configuration/secrets and update them
+before the main container gets it. Then use sidecars to continue to watch for
+changes and perform the updates and push to the main container. This requires
+two separate code paths today. Perhaps the same sidecar could be used for both
+cases.
+
+### Goals
+
+This proposal aims to:
+- make containers implementing the sidecar pattern first class citizens inside a
+  Pod
+- solve a Job completion issue when sidecars should run continuously
+- allow mixing initContainers and sidecars for a choreographed startup sequence
+- allow easy injection of sidecar containers in any Pod
+
+### Non-Goals
+
+This proposal doesn't aim to:
+- support arbitrary dependencies graphs between containers
+- act as a security control to enforce that pod containers only run while the
+  sidecar is healthy. Restart of sidecar containers is a best effort
+- (alpha) support containers termination ordering
+
+## Proposal
+
+The proposal is to introduce a `restartPolicy` field to Init containers and use
+it to indicate that an initContainer is a sidecar container. The only supported
+value proposed in this KEP is “Always”. Kubelet will start these containers in
+the order with other Init containers, but instead of waiting for its completion,
+it will wait for the container startup completion (startup probe succeeded or if
+no startup probe defined, postStart completed).
+
+These containers will not block Pod completion - if all regular containers
+complete, sidecar containers will be terminated.
+
+Those containers will also be restarted even when the Pod restartPolicy is Never
+or OnFailure. Furthermore, sidecar containers will be restarted even during Pod
+termination.
+
+In order to minimize OOM kills of sidecar containers, the OOM adjustment for
+these containers will match or exceed the OOM score adjustment of regular
+containers in the Pod. This intent to solve the issue
+https://github.com/kubernetes/kubernetes/issues/111356 
+
+As part of this KEP we also will be enabling for sidecar containers: PreStart
```

Comment:
Is this supposed to be a list?  Kind of hard to read, maybe use `*` for each bullet?

Reply:
sorry. Fixed